bpf/process/types/basic.h

```diff
@@ -1436,73 +1436,64 @@ struct {
-	__uint(max_entries, MAX_SELECTORS);
-	__uint(key_size, sizeof(u32)); /* selector id */
+	__uint(type, BPF_MAP_TYPE_ARRAY_OF_MAPS);
+	__uint(max_entries, MAX_SELECTORS); // only one matchBinaries per selector
+	__uint(key_size, sizeof(__u32));
 	__array(
 		values, struct {
 			__uint(type, BPF_MAP_TYPE_HASH);
-			__uint(max_entries, 256);
-			__type(key, __u32);
-			__type(value, __u32);
+			__uint(max_entries, 1);
+			__type(key, __u8[MATCH_BINARIES_PATH_MAX_LENGTH]);
+			__type(value, __u8);
 		});
-} sel_names_map SEC(".maps");
-
-static inline __attribute__((always_inline)) int match_binaries(void *sel_names, __u32 selidx)
-{
-	void *binaries_map;
-	struct execve_map_value *execve;
-	__u32 *op, max = 0xffffffff; // UINT32_MAX
-	__u32 ppid, bin_key, *bin_val;
-	bool walker = 0;
-
-	// if binaries_map is NULL for the specific selidx, this
-	// means that the specific selector does not contain any
-	// matchBinaries actions. So we just proceed.
-	binaries_map = map_lookup_elem(sel_names, &selidx);
-	if (binaries_map) {
-		op = map_lookup_elem(binaries_map, &max);
-		if (op) {
-			execve = event_find_curr(&ppid, &walker);
-			if (!execve)
-				return 0;
+} tg_mb_paths SEC(".maps");
 
-			bin_key = execve->binary;
-			bin_val = map_lookup_elem(binaries_map, &bin_key);
+static inline __attribute__((always_inline)) int match_binaries(__u32 selidx)
+{
+	struct execve_map_value *current;
+	__u32 ppid;
+	bool walker, match = 0;
+	void *path_map;
+	__u8 *found_key;
 
-			/*
-			 * The following things may happen:
-			 * binary is not part of names_map, execve_map->binary will be `0` and `bin_val` will always be `0`
-			 * binary is part of `names_map`:
-			 *  if binary is not part of this selector, bin_val will be`0`
-			 *  if binary is part of this selector: `bin_val will be `!0`
-			 */
-			if (*op == op_filter_in) {
-				if (!bin_val)
-					return 0;
-			} else if (*op == op_filter_notin) {
-				if (bin_val)
-					return 0;
-			}
+	struct match_binaries_sel_opts *selector_options;
+
+	// retrieve the selector_options for the matchBinaries, if it's NULL it
+	// means there is not matchBinaries in this selector.
+	selector_options = map_lookup_elem(&tg_mb_sel_opts, &selidx);
+	if (selector_options) {
+		if (selector_options->op == op_filter_none)
+			return 1; // matchBinaries selector is empty <=> match
+
+		current = event_find_curr(&ppid, &walker);
+		if (!current) {
+			// this should not happen, it means that the process was missed when
+			// scanning /proc for process that started before and after tetragon
+			return 0;
+		}
+		if (current->bin.path_length < 0) {
+			// something wrong happened when copying the filename to execve_map
+			return 0;
 		}
+
+		switch (selector_options->op) {
+		case op_filter_in:
+		case op_filter_notin:
+			path_map = map_lookup_elem(&tg_mb_paths, &selidx);
+			if (!path_map)
+				return 0;
+			found_key = map_lookup_elem(path_map, current->bin.path);
+			match = !!found_key;
+			break;
```

Comment:
The way that the code is now is confusing, because the `break` has no effect.

The other question that comes to mind when reading the code is can `->op` have a different value from op_filter_in and op_filter_notin. My guess is no, because we are setting the op from user-space. So _technically_ the switch statement is never used.

My guess is that we would add cases for prefix later.

My suggestion would be to add a 
```
default:
     /* should not happen */
     return 0;
```

In the switch statement to make the code less confusing.

Reply:
Ah yes this is a leftover from this https://github.com/cilium/tetragon/pull/1732/commits/e4c545d0009ae3496147dcd1d13741e6d52218bd for the followup PR, I can adjust!